cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionGoal.java

```diff
@@ -121,9 +118,23 @@ public String name() {
   @Override
   protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
     super.initGoalState(clusterModel, optimizationOptions);
-    for (Broker broker : clusterModel.brokers()) {
-      BrokerAndSortedReplicas bas = new BrokerAndSortedReplicas(broker, broker.replicaComparator(DISK));
-      _brokerAndReplicasMap.put(broker.id(), bas);
+    clusterModel.trackSortedReplicas(name(),
+                                     ReplicaSortFunctionFactory.prioritizeImmigrants(),
+                                     ReplicaSortFunctionFactory.sortByMetricGroupValue(DISK.name()));
+  }
+
+  /**
+   * Update goal state after one round of self-healing / rebalance.
+   * @param clusterModel The state of the cluster.
+   * @param excludedTopics The topics that should be excluded from the optimization proposal.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, Set<String> excludedTopics)
+      throws OptimizationFailureException {
+    super.updateGoalState(clusterModel, excludedTopics);
+    // Clean up memory usage.
+    if (_finished) {
+      clusterModel.untrackSortedReplicas(name());
```

Comment:
[This also applies to the other goals that use `ClusterModel#untrackSortedReplicas(String)`]
Shouldn't we call `ClusterModel#untrackSortedReplicas(String)` when the goal throws an exception -- e.g. `OptimizationFailureException`?